sdk/rust/oak/src/node.rs

```diff
@@ -0,0 +1,84 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::OakStatus;
+use oak_abi::{label::Label, proto::oak::application::NodeConfiguration};
+pub use oak_io::{Decodable, Encodable, Receiver, Sender};
+
+/// Convenience struct for building Nodes.
+pub struct NodeBuilder {
+    config: NodeConfiguration,
+    label: Option<Label>,
+}
+
+// TODO(#1584): Send init messages as a part of building process.
+impl NodeBuilder {
+    pub fn new(config: &NodeConfiguration) -> Self {
+        Self {
+            config: config.clone(),
+            label: None,
+        }
+    }
+
+    /// Set `label` for both newly created Node and a corresponding initial channel to this Node.
+    pub fn label<'a>(&'a mut self, label: &Label) -> &'a mut Self {
+        self.label = Some(label.clone());
+        self
+    }
+
+    /// Build Node using provided [`NodeBuilder::config`] and [`NodeBuilder::label`].
+    /// If [`NodeBuilder::label`] is `None` - uses [`Label::public_untrusted`].
+    /// Returns [`Sender`] for the initial channel to the newly created Node.
+    pub fn build<T: Encodable + Decodable>(&self) -> Result<Sender<T>, OakStatus> {
+        let (sender, receiver) = match &self.label {
+            Some(label) => {
+                let (sender, receiver) = crate::io::channel_create_with_label::<T>(label)?;
+                crate::node_create_with_label(&self.config, label, receiver.handle)?;
+                (sender, receiver)
+            },
```

Comment:
Done
I think people will still use unlabeled nodes a lot, because they are not technically unlabeled - they are labeled with their hashes and optionally signatures

Reply:
What do you mean that they are labelled with their hashes? AFAICT they are labelled as public, which is not the same.

Reply:
Aren't all nodes labeled with hash labels?

I know that they have hash labels as their privileges, but don't they also have hash labels and their labels (so they could read data they they created previously) ?

Reply:
I don't think they have any label other than public unless the creator assigned it. An application that expects to rely on wasm signatures or anything like that would still to assign those labels correctly.

Reply:
Do you have an example for the pattern you are referring to? @ipetr0v maybe we do have this functionality but I just forgot

Reply:
The same way we automatically assign Node privileges:
https://github.com/project-oak/oak/blob/c3b500eb9c54e84f7d88f079c1c14f33e995687a/oak_runtime/src/node/wasm/mod.rs#L983-L985

We can also assign Node labels (it's hash label).

Reply:
The privilege needs to be assigned in a trustworthy way by the runtime, therefore it cannot be exposed. Labels are assigned by the caller node, the meaning of a node confidentiality label is the confidentiality level at which the node may read data. Assigning implicit labels would only work for some very narrow cases, but in general it is not how labels are meant to be used in IFC.

Reply:
Does it mean that I can accidentally (or intentionally) assign a wrong hash label to a node?

Reply:
There is no right or wrong label from an IFC point of view, it can assign any label it wants, but of course that may make it impossible to then send data back to the client or to the backend.

Reply:
It is perfectly fine to create a node with top-secret confidentiality label, there is nothing that violates IFC in that. Of course it is not particularly useful if we expect that node to emit data later on, but that's a different story.